tls/s2n_resume.c

```diff
@@ -769,187 +769,147 @@ struct s2n_ticket_key *s2n_find_ticket_key(struct s2n_config *config, const uint
     return NULL;
 }
 
-int s2n_encrypt_session_ticket(struct s2n_connection *conn, struct s2n_stuffer *to)
+S2N_RESULT s2n_resume_encrypt_session_ticket(struct s2n_connection *conn, struct s2n_stuffer *to)
 {
-    struct s2n_ticket_key *key = NULL;
-    struct s2n_session_key aes_ticket_key = { 0 };
-    struct s2n_blob aes_key_blob = { 0 };
-
-    uint8_t iv_data[S2N_TLS_GCM_IV_LEN] = { 0 };
-    struct s2n_blob iv = { 0 };
-    POSIX_GUARD(s2n_blob_init(&iv, iv_data, sizeof(iv_data)));
-
-    uint8_t aad_data[S2N_TICKET_AAD_LEN] = { 0 };
-    struct s2n_blob aad_blob = { 0 };
-    POSIX_GUARD(s2n_blob_init(&aad_blob, aad_data, sizeof(aad_data)));
-    struct s2n_stuffer aad = { 0 };
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(to);
 
+    struct s2n_ticket_key *key = NULL;
     key = s2n_get_ticket_encrypt_decrypt_key(conn->config);
 
     /* No keys loaded by the user or the keys are either in decrypt-only or expired state */
-    S2N_ERROR_IF(!key, S2N_ERR_NO_TICKET_ENCRYPT_DECRYPT_KEY);
+    RESULT_ENSURE(key != NULL, S2N_ERR_NO_TICKET_ENCRYPT_DECRYPT_KEY);
 
-    POSIX_GUARD(s2n_stuffer_write_bytes(to, key->key_name, S2N_TICKET_KEY_NAME_LEN));
+    /* Initialize AES key */
+    struct s2n_blob aes_key_blob = { 0 };
+    RESULT_GUARD_POSIX(s2n_blob_init(&aes_key_blob, key->aes_key, S2N_AES256_KEY_LEN));
+    DEFER_CLEANUP(struct s2n_session_key aes_ticket_key = { 0 }, s2n_session_key_free);
+    RESULT_GUARD_POSIX(s2n_session_key_alloc(&aes_ticket_key));
+    RESULT_GUARD(s2n_aes256_gcm.init(&aes_ticket_key));
+    RESULT_GUARD(s2n_aes256_gcm.set_encryption_key(&aes_ticket_key, &aes_key_blob));
 
-    POSIX_GUARD_RESULT(s2n_get_public_random_data(&iv));
-    POSIX_GUARD(s2n_stuffer_write(to, &iv));
+    /* Initialize Additional Authenticated Data */
+    uint8_t aad_data[S2N_TICKET_AAD_LEN] = { 0 };
+    struct s2n_blob aad_blob = { 0 };
+    RESULT_GUARD_POSIX(s2n_blob_init(&aad_blob, aad_data, sizeof(aad_data)));
+    struct s2n_stuffer aad = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_init(&aad, &aad_blob));
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(&aad, key->implicit_aad, S2N_TICKET_AAD_IMPLICIT_LEN));
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(&aad, key->key_name, S2N_TICKET_KEY_NAME_LEN));
 
-    POSIX_GUARD(s2n_blob_init(&aes_key_blob, key->aes_key, S2N_AES256_KEY_LEN));
-    POSIX_GUARD(s2n_session_key_alloc(&aes_ticket_key));
-    POSIX_GUARD_RESULT(s2n_aes256_gcm.init(&aes_ticket_key));
-    POSIX_GUARD_RESULT(s2n_aes256_gcm.set_encryption_key(&aes_ticket_key, &aes_key_blob));
+    /* Write key name */
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(to, key->key_name, S2N_TICKET_KEY_NAME_LEN));
```

Comment:
Maybe this would be marginally safer? Same for the other places arrays on the key struct are used. 
```suggestion
    /* Write key name */
    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(to, key->key_name, sizeof(key->key_name)));
```